src/symbols.c

```diff
@@ -2071,102 +2071,114 @@ static TMTag *find_best_goto_tag(GeanyDocument *doc, GPtrArray *tags)
-	type = (definition) ? tm_tag_max_t - forward_types : forward_types;
-	all_tags = tm_workspace_find(name, NULL, type, NULL, old_doc->file_type->lang);
+	all_tags = tm_workspace_find(name, NULL, tm_tag_max_t, NULL, old_doc->file_type->lang);
 
-	/* get rid of global tags */
-	workspace_tags = g_ptr_array_new();
+	/* get rid of global tags and find tag at current line */
+	tags = g_ptr_array_new();
 	foreach_ptr_array(tmtag, i, all_tags)
 	{
 		if (tmtag->file)
-			g_ptr_array_add(workspace_tags, tmtag);
-	}
-
-	/* If there are typedefs of e.g. a struct such as "typedef struct Foo {} Foo;",
-	 * keep just one of the names in the list - when the cursor is on the struct
-	 * name, keep the typename, otherwise keep the struct name. */
-	last_tag = NULL;
-	filtered_tags = g_ptr_array_new();
-	foreach_ptr_array(tmtag, i, workspace_tags)
-	{
-		if (last_tag != NULL && last_tag->file == tmtag->file &&
-			last_tag->type != tm_tag_typedef_t && tmtag->type == tm_tag_typedef_t)
 		{
-			if (last_tag->line == current_line && filtered_tags->len > 0)
-				/* if cursor on struct, replace struct with the typedef */
-				filtered_tags->pdata[filtered_tags->len-1] = tmtag;
-			/* if cursor anywhere else, use struct (already added) and discard typedef */
+			g_ptr_array_add(tags, tmtag);
+			if (tmtag->file == old_doc->tm_file && tmtag->line == current_line)
+				current_tag = tmtag;
 		}
-		else
-			g_ptr_array_add(filtered_tags, tmtag);
+	}
+
+	if (current_tag)
+		/* swap definition/declaration search */
+		definition = current_tag->type & forward_types;
 
-		last_tag = tmtag;
+	filtered_tags = filter_tags(tags, current_tag, definition);
+	if (current_tag && filtered_tags->len == 0)
+	{
+		/* if we previously swapped definition/declaration search and didn't
+		 * find anything, try again with the opposite type */
+		g_ptr_array_free(filtered_tags, TRUE);
+		filtered_tags = filter_tags(tags, current_tag, !definition);
 	}
-	g_ptr_array_free(workspace_tags, TRUE);
-	workspace_tags = filtered_tags;
+	g_ptr_array_free(tags, TRUE);
+	tags = filtered_tags;
 
-	if (workspace_tags->len == 1)
+	if (tags->len == 1)
 	{
 		GeanyDocument *new_doc;
 
-		tmtag = workspace_tags->pdata[0];
-		new_doc = document_find_by_real_path(
-			tmtag->file->file_name);
+		tmtag = tags->pdata[0];
+		new_doc = document_find_by_real_path(tmtag->file->file_name);
 
-		if (new_doc)
-		{
-			/* If we are already on the tag line, swap definition/declaration */
-			if (new_doc == old_doc && tmtag->line == current_line)
-			{
-				if (goto_tag(name, !definition))
-					found = TRUE;
-			}
-		}
-		else
-		{
+		if (!new_doc)
 			/* not found in opened document, should open */
 			new_doc = document_open_file(tmtag->file->file_name, FALSE, NULL, NULL);
-		}
 
-		if (!found && navqueue_goto_line(old_doc, new_doc, tmtag->line))
-			found = TRUE;
+		navqueue_goto_line(old_doc, new_doc, tmtag->line);
 	}
-	else if (workspace_tags->len > 1)
+	else if (tags->len > 1)
```

Comment:
Just renamed workspace_tags to tags because "workspace_tags" was a bit misleading now. No real change below this line.